packages/core-app-api/src/routing/useNavigateUrl.tsx

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { configApiRef, useApi } from '@backstage/core-plugin-api';
+import { useCallback } from 'react';
+import { useNavigate } from 'react-router-dom';
+
+/**
+ * Resolve a URL to a relative URL given a base URL that may or may not include subpaths.
+ * @param url - URL to parse into a relative url based on the baseUrl.
+ * @param baseUrl - Application base url, where the application is currently hosted.
+ * @returns relative path without any subpaths from website config.
+ */
+export function resolveUrlToRelative(url: string, baseUrl: string) {
+  const parsedAppUrl = new URL(baseUrl);
+  const appUrlPath = `${parsedAppUrl.origin}${parsedAppUrl.pathname.replace(
+    /\/$/,
+    '',
+  )}`;
+
+  const relativeUrl = url
+    .replace(appUrlPath, '')
+    // Remove any leading and trailing slashes.
+    .replace(/\/+$/, '')
+    .replace(/^\/+/, '');
+  const parsedUrl = new URL(`http://localhost/${relativeUrl}`);
+  return `${parsedUrl.pathname}${parsedUrl.search}${parsedUrl.hash}`;
+}
+
+/**
+ * A helper hook that allows for full internal website urls to be processed through the navigate
+ *  hook provided by `react-router-dom`.
+ *
+ * NOTE: This does not support routing to external URLs. That should be done with a `Link` or `a`
+ *  element instead, or just `window.location.href`.
+ *
+ * @returns Navigation function that is a wrapper over `react-router-dom`'s
+ *  to support passing full URLs for navigation.
+ *
+ * @public
+ */
+export function useNavigateUrl() {
```

Comment:
A lot of the functionality baked in here is already implemented by the routing system. We've discussed this need before and what I suggested back then was that we add a `appRootRouteRef` that references the root of the app. I think that's the leanest way to go about this that ships this need with the fewest API additions. From what I can tell it'll work to solve this use-case. We'd export that from `@backstage/core-plugin-api`.

What do you think?

Reply:
I agree that this hook isn't needed and would fit better in the existing routing API but I'm not completely following. I took a look at the `@backstage/core-plugin-api` `routing` folder and it seems relevant but not sure how just adding `appRootRouteRef` would solve this issue. It looks like most of the existing routing is meant to translate a `RouteRef` into its corresponding URL not the other way around.

Reply:
@Rugvip wanted to follow up on this, can you explain a little bit more your thinking here?

Reply:
My thinking is that you'd be able to do this:

```tsx
import { appRootRouteRef, useRouteRef } from `@backstage/core-plugin-api`;

function MyComponent() {
  const appRootLink = useRouteRef(appRootRouteRef);

  return <a href={appRootLink()}>Link to app root</a>
}
```

The `appRootRouteRef` would simply be bound to `/` in the app, which the routing system would then resolve with the appropriate base URL.

I see what you mean though, as what you need is to be able to remove that `appRootLink()` from an existing URL. Tbh I feel this is too specific of a need to add to the core APIs. The `appRootRouteRef` does take you halfway though, it avoids needing to read `app.baseUrl` from config, which might not always be correct, the existing logic for that in the app is [here](https://github.com/backstage/backstage/blob/b9828c3c8dcdd72329f190cad801cd0388999b34/packages/core-app-api/src/app/AppRouter.tsx#L37-L59). With that in place I think the rest of the logic should be contained within TechDocs for now. Alternatively we skip the `appRootRouteRef` and duplicate the config reading logic in TechDocs too, but it's a lil bit more fragile.